### PR TITLE
vmm: add configuration for network offloading features

### LINKED ISSUE #4792: Add new features to make TSO, UFO, CHECKSUM, etc. configurable
The host does not support offload capability at the moment, we need to turn off the offload switch of the vm NIC.

If CH can make the TSO, UFO, CHECKSUM features configurable, and turn them off in the default configuration it will make the CH network easier to use.

### PR DESCRIPTION
**Pull request by zhuangel**
Add new configuration for offloading features, including Checksum/TSO/UFO, and set these offloading features as disabled by default.

Fixes: #4792.

Signed-off-by: Yong He <alexyonghe@tencent.com>

### CODE REVIEW

**Review by rbradford — COMMENTED** [review]
Thank you for your creating this PR. Please could you name the variables, datastructure members and CLI option with an `offload_` prefix.

**Review by sboeuf — COMMENTED** [review]
Do we really want/need the granularity that is provided by this patch? We could have something slightly simpler by having a `offload_features=on|off`.

**Comment by rbradford** [comment]
> Do we really want/need the granularity that is provided by this patch? We could have something slightly simpler by having a `offload_features=on|off`.

Unfortunately I think the higher granularity is required. There are various search results about why you might want to turn this functionality off but many of those mailing list posts are very old now.

**Comment by zhuangel** [comment]
> 

Thanks, I have updated the PR.

**Comment by zhuangel** [comment]
> > Do we really want/need the granularity that is provided by this patch? We could have something slightly simpler by having a `offload_features=on|off`.
> 
> Unfortunately I think the higher granularity is required. There are various search results about why you might want to turn this functionality off but many of those mailing list posts are very old now.

Sorry, my understanding of this problem is consistent with what is described in https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4792. I will try to get an update on the issue.

**Review by sboeuf — APPROVED** [review]
Looks good!

**Comment by rbradford** [comment]
@zhuangel You need to fix some of the issues found by GitHub actions

**Comment by zhuangel** [comment]
> @zhuangel You need to fix some of the issues found by GitHub actions

Thanks, I have fixed 2 format issue and 1 fuzz issue, and I will check the workflow results.

**Comment by zhuangel** [comment]
> > > Do we really want/need the granularity that is provided by this patch? We could have something slightly simpler by having a `offload_features=on|off`.
> > 
> > 
> > Unfortunately I think the higher granularity is required. There are various search results about why you might want to turn this functionality off but many of those mailing list posts are very old now.
> 
> Sorry, my understanding of this problem is consistent with what is described in #4792. I will try to get an update on the issue.

Some update on this, the network architecture is a little complicated to me, for short there maybe some environment has no offload feature support, so this PR just want some configurations so that we could disable offload features when needed, thanks.

**Review by likebreath — COMMENTED** [review]
The commit message is misleading and contradicts with what the PR does.  I guess you meant to say: "... these offloading features as `*enabled*` by default "

**Review by rbradford — COMMENTED** [review]
I think this change might need an integration test. I think we already have some testing around ethtool for the virtio-net control queue that allows the enabling/disabling of these offloads at runtime.

**Comment by zhuangel** [comment]
> The commit message is misleading and contradicts with what the PR does. I guess you meant to say: "... these offloading features as `*enabled*` by default "

Thanks, I will update the commit log.

File: virtio-devices/src/net.rs [hunk 1]
```diff
@@ -462,25 +465,32 @@ impl Net {
...
-                    | 1 << VIRTIO_RING_F_EVENT_IDX
-                    | 1 << VIRTIO_F_VERSION_1;
+                let mut avail_features =
+                    1 << VIRTIO_NET_F_MTU | 1 << VIRTIO_RING_F_EVENT_IDX | 1 << VIRTIO_F_VERSION_1;
 
                 if iommu {
                     avail_features |= 1u64 << VIRTIO_F_IOMMU_PLATFORM;
                 }
 
+                if offload_csum {
```
Comment (rbradford) [inline virtio-devices/src/net.rs:475]: I had to look at the specification to understand this interdependency. so a comment would be useful here.

File: virtio-devices/src/net.rs [hunk 2]
```diff
@@ -462,25 +465,32 @@ impl Net {
...
+                            | 1 << VIRTIO_NET_F_HOST_TSO6
+                            | 1 << VIRTIO_NET_F_GUEST_ECN
+                            | 1 << VIRTIO_NET_F_GUEST_TSO4
+                            | 1 << VIRTIO_NET_F_GUEST_TSO6;
+                    }
+
+                    if offload_ufo {
+                        avail_features |= 1 << VIRTIO_NET_F_HOST_UFO | 1 << VIRTIO_NET_F_GUEST_UFO;
+                    }
+                }
```
Comment (rbradford) [inline virtio-devices/src/net.rs:492]: Due to these interdependencies you need to make `NetConfig::validate()` check that the user isn't providing an invalid configuration.

**Review by rbradford — CHANGES_REQUESTED** [review]
I think this needs an integration test as well as validation and parsing unit tests (in `config.rs`)

File: virtio-devices/src/net.rs (hunk 1 above)
Comment (zhuangel) [reply virtio-devices/src/net.rs:476]: Thanks, fixed.

File: virtio-devices/src/net.rs (hunk 2 above)
Comment (zhuangel) [reply virtio-devices/src/net.rs:493]: Thanks, fixed and test case added.

**Comment by zhuangel** [comment]
> I think this needs an integration test as well as validation and parsing unit tests (in `config.rs`)

The integration is a little trick, this may need to force disable host (tap) offloading feature, if host create a tap with offloading feature enabled, this may cause the host to vm network issue.

I didn't have a complete case for this now.

**Comment by rbradford** [comment]
> > I think this needs an integration test as well as validation and parsing unit tests (in `config.rs`)
> 
> The integration is a little trick, this may need to force disable host (tap) offloading feature, if host create a tap with offloading feature enabled, this may cause the host to vm network issue.
> 
> I didn't have a complete case for this now.

Okay, don't worry about it. We need the networking for the VM for the integration testing to work :-)

**Review by rbradford — COMMENTED** [review]
Very close. Just some small nits.

File: vmm/src/config.rs [hunk 3]
```diff
@@ -205,6 +207,10 @@ impl fmt::Display for ValidationError {
                 "Number of queues to virtio_net does not match the number of input FDs"
             ),
             VnetReservedFd => write!(f, "Reserved fd number (<= 2)"),
+            NoHarewareChecksumOffload => write!(
```
Comment (rbradford) [inline vmm/src/config.rs:210]: There is also a typo in the name of the error enum value.

Comment (zhuangel) [reply vmm/src/config.rs]: Thanks, fixed.

File: virtio-devices/src/vhost_user/net.rs [hunk 4]
```diff
@@ -140,6 +133,24 @@ impl Net {
                 avail_features |= 1u64 << VIRTIO_NET_F_MTU;
             }
 
+            // Configure TSO/UFO features when hardward checksum offload is enabled.
```
[the remark this answers is not in the document]
Comment (zhuangel) [reply virtio-devices/src/vhost_user/net.rs]: Thanks, fixed.

File: vmm/src/config.rs [hunk 5]
```diff
@@ -205,6 +207,10 @@ impl fmt::Display for ValidationError {
...
+                f,
+                "Enable offload_csum before enable offload_tso and offload_ufo"
```
[the remark this answers is not in the document]
Comment (zhuangel) [reply vmm/src/config.rs]: Thanks, fixed.

File: virtio-devices/src/net.rs [hunk 6]
```diff
@@ -462,25 +465,33 @@ impl Net {
...
-                    | 1 << VIRTIO_RING_F_EVENT_IDX
-                    | 1 << VIRTIO_F_VERSION_1;
+                let mut avail_features =
+                    1 << VIRTIO_NET_F_MTU | 1 << VIRTIO_RING_F_EVENT_IDX | 1 << VIRTIO_F_VERSION_1;
 
                 if iommu {
                     avail_features |= 1u64 << VIRTIO_F_IOMMU_PLATFORM;
                 }
 
+                // Configure TSO/UFO features when hardward checksum offload is enabled.
```
[the remark this answers is not in the document]
Comment (zhuangel) [reply virtio-devices/src/net.rs]: Thanks, fixed

**Review by rbradford — APPROVED** [review]

**Comment by rbradford** [comment]
I pushed a small change to the commit message to add the word network in the short summary.

**Comment by zhuangel** [comment]
> I pushed a small change to the commit message to add the word network in the short summary.

Thanks for your help and review !

**Comment by wfly1998** [comment]
Hey, is it necessary to split those configs into two parts, like `offload_host_xxx` and `offload_guest_xxx`?  I think there may be cases where only host features or only guest features are used.

**Pull request merged**